test/Microsoft.Azure.Devices.Edge.Test.Common/config/BaseModuleConfigBuilder.cs

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Test.Common.Config
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.ObjectModel;
+    using System.Linq;
+    using Microsoft.Azure.Devices.Edge.Util;
+
+    public abstract class BaseModuleConfigBuilder : IModuleConfigBuilder
+    {
+        IDictionary<string, object> DesiredProperties { get; }
+
+        IDictionary<string, string> Env { get; }
+
+        protected IDictionary<string, object> Deployment { get; }
+
+        protected IDictionary<string, object> Settings { get; }
+
+        public string Name { get; }
+
+        protected BaseModuleConfigBuilder(string name, string image)
+        {
+            this.Deployment = new Dictionary<string, object>()
+            {
+                ["type"] = "docker"
+            };
+            this.DesiredProperties = new Dictionary<string, object>();
+            this.Env = new Dictionary<string, string>();
+            this.Name = name;
+            this.Settings = new Dictionary<string, object>()
+            {
+                ["image"] = image
+            };
+        }
+
+        public IModuleConfigBuilder WithEnvironment(IEnumerable<(string, string)> env)
+        {
+            foreach ((string key, string value) in env)
+            {
+                this.Env[key] = value; // for duplicate keys, last save wins!
+            }
+
+            return this;
+        }
+
+        public IModuleConfigBuilder WithProxy(Option<Uri> proxy, Protocol protocol)
```

Comment:
Nit: Fine for test code, but might be cleaner to have a 
WithProxy(Uri proxy)
WithProtocol(Protocol)
and have the logic to update protocol in Build.

Reply:
I felt like I needed to add that `protocol` parameter since the proxy setting has an impact on upstream protocol, but I didn't like it. In thinking about it, maybe it would be better to just have `WithProxy(Uri proxy)`. If the caller also specifies `WithEnvironment(new[] { ("UpstreamProtocol", p) })` and `p` is not compatible with the proxy (or if they specify an incompatible "UpstreamProtocol" first and then add the proxy), then just throw instead of trying to convert it. The caller can fix the bug in their test.

Reply:
Updated.